### ISSUE TITLE
Storage component only charge by anticipation of a discharge

### ISSUE THREAD

**Issue by JeromeBaltha** [cenaero-enb/h2cs-design#14]
**Note: for support questions and enhancement suggestions please use the [VPPopt gitter community](https://gitter.im/VPPopt/VPPopt-Support)**. This repository's issues are reserved for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] bug report
  - [ ] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

* **Do you want to request a *feature* or report a *bug*?**
Report a *bug*

* **What is the current behavior?**

A simple energetic system, composed of:
- 1 electricity bus with an excess cost exaggerated of 1000000 €/kW
- 1 access to electrical network with a variable cost of 100 €/kW
-  2 productions (non-normalized profiles)
- 1 consumption (non-normalized profile) (production generally higher than consumption)
- 1 simple storage component of 10000 kWh (no losses, full efficiency, variable input cost = 1 €/kW, variable output cost = 1€/kW)

Representation of the system:
![Figure_1](https://user-images.githubusercontent.com/100592710/172363027-81e879f4-a899-40f8-bc95-41d7586a9228.png)

The system is solved by the CBC solver.
It's a dispatch optimization (no sizing of components)

Excel of inputs available in attached files: 
[scenario.xlsx](https://github.com/cenaero-enb/h2cs-design/files/8852427/scenario.xlsx)

The current behavior is that the solver only organize input flows in the storage when he knows the storage will be useful for complementing the production to satisfy the consumption. Indeed, if at one timestep, the total production is lower than the total consumption, it's cheaper to discharge the battery than buying electricity to the network. From this timestep, the battery is charged by the previous timestep to satisfy only the quantity of energy that will be discharged. If the previous timestep is not sufficient, the battery will be charged also in a new previous timestep and again and again, until the moment discharged quantity is satisfied. All this knowing that it's normally a lot cheaper to charge the battery than to re-inject electricity in excess.

A visual explication is available in the attached screenshot (the green cases are when the production is bigger than the consumption):
![Capture](https://user-images.githubusercontent.com/100592710/172362412-1c103f6b-64a9-4e1a-9f6e-53ca8e65c69c.PNG)

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

Input excel available in attached file.

* **What is the expected behavior?**

The behavior that would be expected is that the storage begins to charge despite there is no discharging in a near horizon. And only once the storage is full or the actual maximum possible flow is reached, the rest of the production can be re-injected to the network at a higher cost.

* **What is the motivation / use case for changing the behavior?**

If the storage is only charged in anticipation of a discharge, it's not a correct behavior. It can also be problematic for the EMS because it can't predict the future behaviour.

* **Please tell us about your environment:**
  
  - VPPopt Version: 0.1.0
  - Operating System: Windows 64-bit
  - Python version: 3.9

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

The same optimization with the GLPK solver gives the same results.
So maybe it's linked to the constraints written in the model by oemof solph.
I checked and the variable input and output costs of the flows of the storage are written in the objective function. So the optimization is not at the minimal cost of operation after optimization because at each timestep the optimizer should charge the storage by priority because this flow is cheaper.

Suggestions to fix:
- Parameters to give to the solver ?
- Add priority to the storage flows ?

**Comment by longlevan** [comment]
I see another problem but I have to check for that and comme to you later for the explanation. Indeed, by default Storage model created by vppopt (or oemof.solph) has a constraint (set by an argument in Storage model setting parameter) which is _**the state of charge in the last time step is equal to initial value in time step zero.**_  As you performed an optimization just for few days without any loss on battery capacity (in Excel file capacity loss = 0, efficiency inflow = efficiency outflow) and the energy production is almost always higher than the consumption, the solver decided to find a way to respect this constraint first and find then find an optimal solution then.

As you can see from your out.xls, if solver decide to charge the battery when there is excess electricity and have a state of charge of battery for example 20 000 kWh, then there are no loss, no discharge (as production > consumption), the state of charge of battery at final timestep can never equal to this at initial timestep.

**Comment by JeromeBaltha** [comment]
OK i get it ! 

I saw this constraint when i was exploring the scripts, but I always thought it was not active but now i found where it was declared.

And that's the reason why at the first or last timestep there is sometimes a sudden massive charge or discharge to respect this constraint. I suppose.

However, i don't understand exactly what is the goal of a such constraint ?

Thank you very much.
